enterprise/cypher/spec-suite-tools/src/main/scala/cypher/feature/parser/SideEffects.scala

```diff
@@ -0,0 +1,155 @@
+      case (v, (ADDED_LABELS, index)) =>
+        v.copy(labelsCreated = values.get(index).toInt)
+      case (v, (DELETED_LABELS, index)) =>
+        v.copy(labelsDeleted = values.get(index).toInt)
+      case (v, (ADDED_PROPERTIES, index)) =>
+        v.copy(propsCreated = values.get(index).toInt)
+      case (v, (DELETED_PROPERTIES, index)) =>
+        v.copy(propsDeleted = values.get(index).toInt)
+      case (sideEffect, _) => throw InvalidFeatureFormatException(
+        s"Invalid side effect: $sideEffect. Valid ones are: ${ALL.mkString(",")}")
+    }
+
+    expected
+  }
+
+  case class Values(nodesCreated: Int = 0,
+                    nodesDeleted: Int = 0,
+                    relsCreated: Int = 0,
+                    relsDeleted: Int = 0,
+                    labelsCreated: Int = 0,
+                    labelsDeleted: Int = 0,
+                    propsCreated: Int = 0,
+                    propsDeleted: Int = 0) {
+    override def toString = {
+      s"""Values(
+          |  +nodes:  $nodesCreated
+          |  -nodes:  $nodesDeleted
+          |  +rels:   $relsCreated
+          |  -rels:   $relsDeleted
+          |  +labels: $labelsCreated
+          |  -labels: $labelsDeleted
+          |  +props:  $propsCreated
+          |  -props:  $propsDeleted)""".stripMargin
+    }
+  }
+
+  case class State(nodes: Set[Long] = Set.empty,
+                   rels: Set[Long] = Set.empty,
+                   labels: Set[String] = Set.empty,
+                   props: Set[(Entity, String, AnyRef)] = Set.empty) {
+
+    /**
+      * Computes the difference in between this state and a later state (the argument).
+      * The difference is a set of side effects in the form of a Values instance.
+      *
+      * @param later the later state to compare against.
+      * @return the side effect difference, as a Values instance.
+      */
+    def diff(later: State): Values = {
+      val nodesCreated = (later.nodes diff nodes).size
+      val nodesDeleted = (nodes diff later.nodes).size
+      val relsCreated = (later.rels diff rels).size
+      val relsDeleted = (rels diff later.rels).size
+      val labelsCreated = (later.labels diff labels).size
+      val labelsDeleted = (labels diff later.labels).size
+      val propsCreated = (later.props diff props).size
+      val propsDeleted = (props diff later.props).size
+
+      Values(
+        nodesCreated, nodesDeleted,
+        relsCreated, relsDeleted,
+        labelsCreated, labelsDeleted,
+        propsCreated, propsDeleted
+      )
+    }
+  }
+
+  private val prefix = "CYPHER 3.3 runtime=interpreted"
+
+  private val nodesQuery =
+    s"""$prefix MATCH (n) RETURN id(n) AS node"""
+
+  private val relsQuery =
+    s"""$prefix MATCH ()-[r]->() RETURN id(r) AS rel"""
+
+  private val labelsQuery =
+    s"""$prefix MATCH (n)
+       |UNWIND labels(n) AS label
+       |RETURN DISTINCT label""".stripMargin
+
+  private val propsQuery =
+    s"""$prefix MATCH (n)
```

Comment:
Would it not be easier to `CALL db.propertyKeys()` ?

Reply:
Although I suspect that will not show deleted properties, right?

Reply:
These queries are part of the standard TCK definitions. See https://github.com/opencypher/openCypher/tree/master/tck#observability-of-side-effects

The idea behind that is to formulate the TCK in terms of standard Cypher only. Currently, there are no procedures that are part of the standard.